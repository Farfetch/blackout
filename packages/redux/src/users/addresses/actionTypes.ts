/**
 * Action type dispatched when the create address request fails.
 */
export const CREATE_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the create address request starts.
 */
export const CREATE_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the create address request succeeds.
 */
export const CREATE_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the fetch address request fails.
 */
export const FETCH_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the fetch address request starts.
 */
export const FETCH_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the fetch address request succeeds.
 */
export const FETCH_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the fetch addresses request fails.
 */
export const FETCH_USER_ADDRESSES_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESSES_FAILURE';
/**
 * Action type dispatched when the fetch addresses request starts.
 */
export const FETCH_USER_ADDRESSES_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESSES_REQUEST';
/**
 * Action type dispatched when the fetch addresses request succeeds.
 */
export const FETCH_USER_ADDRESSES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ADDRESSES_SUCCESS';

/**
 * Action type dispatched when the fetch default contact address request fails.
 */
export const FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_CONTACT_ADDRESS_FAILURE';
/**
 * Action type dispatched when the fetch default contact address request starts.
 */
export const FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_CONTACT_ADDRESS_REQUEST';
/**
 * Action type dispatched when the fetch default contact address request succeeds.
 */
export const FETCH_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the delete address request fails.
 */
export const REMOVE_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/REMOVE_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the delete address request starts.
 */
export const REMOVE_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/REMOVE_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the delete address request succeeds.
 */
export const REMOVE_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when delete default contact address request fails.
 */
export const REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/REMOVE_USER_DEFAULT_CONTACT_ADDRESS_FAILURE';
/**
 * Action type dispatched when delete default contact address request starts.
 */
export const REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/REMOVE_USER_DEFAULT_CONTACT_ADDRESS_REQUEST';
/**
 * Action type dispatched when delete default contact address request succeeds.
 */
export const REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS';

/**
 * Action type dispatched when set default billing address request fails.
 */
export const SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_BILLING_ADDRESS_FAILURE';
/**
 * Action type dispatched when set default billing address request starts.
 */
export const SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_BILLING_ADDRESS_REQUEST';
/**
 * Action type dispatched when set default billing address request succeeds.
 */
export const SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_BILLING_ADDRESS_SUCCESS';

/**
 * Action type dispatched when set default shipping address request fails.
 */
export const SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_SHIPPING_ADDRESS_FAILURE';
/**
 * Action type dispatched when set default shipping address request starts.
 */
export const SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_SHIPPING_ADDRESS_REQUEST';
/**
 * Action type dispatched when set default shipping address request succeeds.
 */
export const SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_SHIPPING_ADDRESS_SUCCESS';

/**
 * Action type dispatched when set default contact address request fails.
 */
export const SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_CONTACT_ADDRESS_FAILURE';
/**
 * Action type dispatched when set default contact address request starts.
 */
export const SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_CONTACT_ADDRESS_REQUEST';
/**
 * Action type dispatched when set default contact address request succeeds.
 */
export const SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_CONTACT_ADDRESS_SUCCESS';

/**
 * Action type dispatched when the update address request fails.
 */
export const UPDATE_USER_ADDRESS_FAILURE =
  '@farfetch/blackout-redux/UPDATE_USER_ADDRESS_FAILURE';
/**
 * Action type dispatched when the update address request starts.
 */
export const UPDATE_USER_ADDRESS_REQUEST =
  '@farfetch/blackout-redux/UPDATE_USER_ADDRESS_REQUEST';
/**
 * Action type dispatched when the update address request succeeds.
 */
export const UPDATE_USER_ADDRESS_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_USER_ADDRESS_SUCCESS';

/**
 * Action type dispatched when resetting the addresses.
 */
export const RESET_USER_ADDRESSES =
  '@farfetch/blackout-redux/RESET_USER_ADDRESSES';
