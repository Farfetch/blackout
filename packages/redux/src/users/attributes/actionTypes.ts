/**
 * Action type dispatched when the fetch user attributes request fails.
 */
export const FETCH_USER_ATTRIBUTES_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTES_FAILURE';
/**
 * Action type dispatched when the fetch user attributes request starts.
 */
export const FETCH_USER_ATTRIBUTES_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTES_REQUEST';
/**
 * Action type dispatched when the fetch user attributes request succeeds.
 */
export const FETCH_USER_ATTRIBUTES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTES_SUCCESS';

/**
 * Action type dispatched when the create user attributes request fails.
 */
export const CREATE_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the create user attributes request starts.
 */
export const CREATE_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the create user attributes request succeeds.
 */
export const CREATE_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the fetch user attribute request fails.
 */
export const FETCH_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the fetch user attribute request starts.
 */
export const FETCH_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the fetch user attribute request succeeds.
 */
export const FETCH_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the set user attribute request fails.
 */
export const SET_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/SET_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the set user attribute request starts.
 */
export const SET_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/SET_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the set user attribute request succeeds.
 */
export const SET_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the update user attribute request fails.
 */
export const UPDATE_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/UPDATE_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the update user attribute request starts.
 */
export const UPDATE_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/UPDATE_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the update user attribute request succeeds.
 */
export const UPDATE_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_USER_ATTRIBUTE_SUCCESS';

/**
 * Action type dispatched when the remove user attribute request fails.
 */
export const REMOVE_USER_ATTRIBUTE_FAILURE =
  '@farfetch/blackout-redux/REMOVE_USER_ATTRIBUTE_FAILURE';
/**
 * Action type dispatched when the remove user attribute request starts.
 */
export const REMOVE_USER_ATTRIBUTE_REQUEST =
  '@farfetch/blackout-redux/REMOVE_USER_ATTRIBUTE_REQUEST';
/**
 * Action type dispatched when the remove user attribute request succeeds.
 */
export const REMOVE_USER_ATTRIBUTE_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_USER_ATTRIBUTE_SUCCESS';
