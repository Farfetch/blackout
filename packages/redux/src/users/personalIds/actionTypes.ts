/**
 * Action type dispatched when the fetch personal ids request fails.
 */
export const FETCH_USER_PERSONAL_IDS_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_PERSONAL_IDS_FAILURE';
/**
 * Action type dispatched when the fetch personal ids request starts.
 */
export const FETCH_USER_PERSONAL_IDS_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_PERSONAL_IDS_REQUEST';
/**
 * Action type dispatched when the fetch personal ids request succeeds.
 */
export const FETCH_USER_PERSONAL_IDS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_PERSONAL_IDS_SUCCESS';

/**
 * Action type dispatched when the create personal id request fails.
 */
export const CREATE_USER_PERSONAL_ID_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_PERSONAL_ID_FAILURE';
/**
 * Action type dispatched when the create personal id request starts.
 */
export const CREATE_USER_PERSONAL_ID_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_PERSONAL_ID_REQUEST';
/**
 * Action type dispatched when the create personal id request succeeds.
 */
export const CREATE_USER_PERSONAL_ID_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_PERSONAL_ID_SUCCESS';

/**
 * Action type dispatched when the fetch default personal id request fails.
 */
export const FETCH_USER_DEFAULT_PERSONAL_ID_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_PERSONAL_ID_FAILURE';
/**
 * Action type dispatched when the fetch default personal id request starts.
 */
export const FETCH_USER_DEFAULT_PERSONAL_ID_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_PERSONAL_ID_REQUEST';
/**
 * Action type dispatched when the fetch default personal id request succeeds.
 */
export const FETCH_USER_DEFAULT_PERSONAL_ID_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_DEFAULT_PERSONAL_ID_SUCCESS';

/**
 * Action type dispatched when the set default personal id request fails.
 */
export const SET_USER_DEFAULT_PERSONAL_ID_FAILURE =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_PERSONAL_ID_FAILURE';
/**
 * Action type dispatched when the set default personal id request starts.
 */
export const SET_USER_DEFAULT_PERSONAL_ID_REQUEST =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_PERSONAL_ID_REQUEST';
/**
 * Action type dispatched when the set default personal id request succeeds.
 */
export const SET_USER_DEFAULT_PERSONAL_ID_SUCCESS =
  '@farfetch/blackout-redux/SET_USER_DEFAULT_PERSONAL_ID_SUCCESS';

/**
 * Action type dispatched when the create personal id image request fails.
 */
export const CREATE_USER_PERSONAL_ID_IMAGE_FAILURE =
  '@farfetch/blackout-redux/CREATE_USER_PERSONAL_ID_IMAGE_FAILURE';
/**
 * Action type dispatched when the create personal id image request starts.
 */
export const CREATE_USER_PERSONAL_ID_IMAGE_REQUEST =
  '@farfetch/blackout-redux/CREATE_USER_PERSONAL_ID_IMAGE_REQUEST';
/**
 * Action type dispatched when the create personal id image request succeeds.
 */
export const CREATE_USER_PERSONAL_ID_IMAGE_SUCCESS =
  '@farfetch/blackout-redux/CREATE_USER_PERSONAL_ID_IMAGE_SUCCESS';

/**
 * Action type dispatched when the fetch personal id request fails.
 */
export const FETCH_USER_PERSONAL_ID_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_PERSONAL_ID_FAILURE';
/**
 * Action type dispatched when the fetch personal id request starts.
 */
export const FETCH_USER_PERSONAL_ID_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_PERSONAL_ID_REQUEST';
/**
 * Action type dispatched when the fetch personal id request succeeds.
 */
export const FETCH_USER_PERSONAL_ID_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_PERSONAL_ID_SUCCESS';

/**
 * Action type dispatched when the update personal id request fails.
 */
export const UPDATE_USER_PERSONAL_ID_FAILURE =
  '@farfetch/blackout-redux/UPDATE_USER_PERSONAL_ID_FAILURE';
/**
 * Action type dispatched when the update personal id request starts.
 */
export const UPDATE_USER_PERSONAL_ID_REQUEST =
  '@farfetch/blackout-redux/UPDATE_USER_PERSONAL_ID_REQUEST';
/**
 * Action type dispatched when the update personal id request succeeds.
 */
export const UPDATE_USER_PERSONAL_ID_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_USER_PERSONAL_ID_SUCCESS';

/**
 * Action type dispatched when the remove personal id request fails.
 */
export const REMOVE_USER_PERSONAL_ID_FAILURE =
  '@farfetch/blackout-redux/REMOVE_USER_PERSONAL_ID_FAILURE';
/**
 * Action type dispatched when the remove personal id request starts.
 */
export const REMOVE_USER_PERSONAL_ID_REQUEST =
  '@farfetch/blackout-redux/REMOVE_USER_PERSONAL_ID_REQUEST';
/**
 * Action type dispatched when the remove personal id request succeeds.
 */
export const REMOVE_USER_PERSONAL_ID_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_USER_PERSONAL_ID_SUCCESS';

/**
 * Action type dispatched when resetting the personal ids.
 */
export const RESET_USER_PERSONAL_IDS =
  '@farfetch/blackout-redux/RESET_USER_PERSONAL_IDS';
