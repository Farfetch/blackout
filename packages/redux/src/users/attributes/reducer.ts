import * as actionTypes from './actionTypes';
import { type AnyAction, combineReducers, type Reducer } from 'redux';
import type * as T from './types';

export const INITIAL_STATE: T.UserAttributesState = {
  result: null,
  error: null,
  isLoading: false,
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: AnyAction,
): T.UserAttributesState['isLoading'] => {
  switch (action.type) {
    case actionTypes.CREATE_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTE_REQUEST:
    case actionTypes.REMOVE_USER_ATTRIBUTE_REQUEST:
    case actionTypes.SET_USER_ATTRIBUTE_REQUEST:
    case actionTypes.UPDATE_USER_ATTRIBUTE_REQUEST:
      return true;
    case actionTypes.CREATE_USER_ATTRIBUTES_FAILURE:
    case actionTypes.CREATE_USER_ATTRIBUTES_SUCCESS:
    case actionTypes.FETCH_USER_ATTRIBUTES_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTES_SUCCESS:
    case actionTypes.FETCH_USER_ATTRIBUTE_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTE_SUCCESS:
    case actionTypes.REMOVE_USER_ATTRIBUTE_FAILURE:
    case actionTypes.REMOVE_USER_ATTRIBUTE_SUCCESS:
    case actionTypes.SET_USER_ATTRIBUTE_FAILURE:
    case actionTypes.SET_USER_ATTRIBUTE_SUCCESS:
    case actionTypes.UPDATE_USER_ATTRIBUTE_FAILURE:
    case actionTypes.UPDATE_USER_ATTRIBUTE_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

type ErrorAction =
  | T.CreateUserAttributesFailureAction
  | T.FetchUserAttributeFailureAction
  | T.FetchUserAttributesFailureAction
  | T.RemoveUserAttributeFailureAction
  | T.SetUserAttributeFailureAction
  | T.UpdateUserAttributeFailureAction;

const error = (
  state = INITIAL_STATE.error,
  action: AnyAction,
): T.UserAttributesState['error'] => {
  switch (action.type) {
    case actionTypes.CREATE_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTES_REQUEST:
    case actionTypes.FETCH_USER_ATTRIBUTE_REQUEST:
    case actionTypes.REMOVE_USER_ATTRIBUTE_REQUEST:
    case actionTypes.SET_USER_ATTRIBUTE_REQUEST:
    case actionTypes.UPDATE_USER_ATTRIBUTE_REQUEST:
      return INITIAL_STATE.error;
    case actionTypes.CREATE_USER_ATTRIBUTES_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTES_FAILURE:
    case actionTypes.FETCH_USER_ATTRIBUTE_FAILURE:
    case actionTypes.REMOVE_USER_ATTRIBUTE_FAILURE:
    case actionTypes.SET_USER_ATTRIBUTE_FAILURE:
    case actionTypes.UPDATE_USER_ATTRIBUTE_FAILURE:
      return (action as ErrorAction).payload.error;
    default:
      return state;
  }
};

const result = (
  state = INITIAL_STATE.result,
  action: AnyAction,
): T.UserAttributesState['result'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_ATTRIBUTES_SUCCESS:
      return (action as T.FetchUserAttributesSuccessAction).payload;
    case actionTypes.FETCH_USER_ATTRIBUTE_SUCCESS:
      let currentResult = state;

      if (!currentResult || !Array.isArray(currentResult)) {
        currentResult = [];
      }

      const actionSafe = action as T.FetchUserAttributeSuccessAction;
      const newResult = [...currentResult];

      const index = currentResult.findIndex(
        attribute => attribute?.id === actionSafe.payload.id,
      );

      if (index !== -1) {
        newResult[index] = actionSafe.payload;
      } else {
        newResult.push(actionSafe.payload);
      }

      return newResult;

    default:
      return state;
  }
};

const attributesReducer: Reducer<T.UserAttributesState> = combineReducers({
  isLoading,
  result,
  error,
});

export default attributesReducer;
