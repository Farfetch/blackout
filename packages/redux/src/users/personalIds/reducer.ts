import * as actionTypes from './actionTypes.js';
import { type AnyAction, combineReducers, type Reducer } from 'redux';
import { createReducerWithResult } from '../../helpers/reducerFactory.js';
import type * as T from './types/index.js';

export const INITIAL_STATE: T.UserPersonalIdsState = {
  result: null,
  error: null,
  isLoading: false,
  defaultPersonalId: {
    result: null,
    error: null,
    isLoading: false,
  },
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: AnyAction,
): T.UserPersonalIdsState['isLoading'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_PERSONAL_IDS_REQUEST:
    case actionTypes.CREATE_USER_PERSONAL_ID_REQUEST:
    case actionTypes.FETCH_USER_DEFAULT_PERSONAL_ID_REQUEST:
    case actionTypes.SET_USER_DEFAULT_PERSONAL_ID_REQUEST:
    case actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_REQUEST:
    case actionTypes.FETCH_USER_PERSONAL_ID_REQUEST:
    case actionTypes.UPDATE_USER_PERSONAL_ID_REQUEST:
    case actionTypes.REMOVE_USER_PERSONAL_ID_REQUEST:
      return true;
    case actionTypes.FETCH_USER_PERSONAL_IDS_FAILURE:
    case actionTypes.FETCH_USER_PERSONAL_IDS_SUCCESS:
    case actionTypes.CREATE_USER_PERSONAL_ID_FAILURE:
    case actionTypes.CREATE_USER_PERSONAL_ID_SUCCESS:
    case actionTypes.FETCH_USER_DEFAULT_PERSONAL_ID_FAILURE:
    case actionTypes.FETCH_USER_DEFAULT_PERSONAL_ID_SUCCESS:
    case actionTypes.SET_USER_DEFAULT_PERSONAL_ID_FAILURE:
    case actionTypes.SET_USER_DEFAULT_PERSONAL_ID_SUCCESS:
    case actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_FAILURE:
    case actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_SUCCESS:
    case actionTypes.FETCH_USER_PERSONAL_ID_FAILURE:
    case actionTypes.FETCH_USER_PERSONAL_ID_SUCCESS:
    case actionTypes.UPDATE_USER_PERSONAL_ID_FAILURE:
    case actionTypes.UPDATE_USER_PERSONAL_ID_SUCCESS:
    case actionTypes.REMOVE_USER_PERSONAL_ID_FAILURE:
    case actionTypes.REMOVE_USER_PERSONAL_ID_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

type ErrorAction =
  | T.FetchUserPersonalIdsFailureAction
  | T.CreateUserPersonalIdFailureAction
  | T.FetchUserDefaultPersonalIdFailureAction
  | T.SetUserDefaultPersonalIdFailureAction;

const error = (
  state = INITIAL_STATE.error,
  action: AnyAction,
): T.UserPersonalIdsState['error'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_PERSONAL_IDS_REQUEST:
    case actionTypes.CREATE_USER_PERSONAL_ID_REQUEST:
    case actionTypes.FETCH_USER_DEFAULT_PERSONAL_ID_REQUEST:
    case actionTypes.SET_USER_DEFAULT_PERSONAL_ID_REQUEST:
    case actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_REQUEST:
    case actionTypes.FETCH_USER_PERSONAL_ID_REQUEST:
    case actionTypes.UPDATE_USER_PERSONAL_ID_REQUEST:
    case actionTypes.REMOVE_USER_PERSONAL_ID_REQUEST:
      return INITIAL_STATE.error;
    case actionTypes.FETCH_USER_PERSONAL_IDS_FAILURE:
    case actionTypes.CREATE_USER_PERSONAL_ID_FAILURE:
    case actionTypes.FETCH_USER_DEFAULT_PERSONAL_ID_FAILURE:
    case actionTypes.SET_USER_DEFAULT_PERSONAL_ID_FAILURE:
    case actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_FAILURE:
    case actionTypes.FETCH_USER_PERSONAL_ID_FAILURE:
    case actionTypes.UPDATE_USER_PERSONAL_ID_FAILURE:
    case actionTypes.REMOVE_USER_PERSONAL_ID_FAILURE:
      return (action as ErrorAction).payload.error;
    default:
      return state;
  }
};

const result = (
  state = INITIAL_STATE.result,
  action: AnyAction,
): T.UserPersonalIdsState['result'] => {
  switch (action.type) {
    case actionTypes.FETCH_USER_PERSONAL_IDS_SUCCESS:
      return (action as T.FetchUserPersonalIdsSuccessAction).payload;
    case actionTypes.FETCH_USER_PERSONAL_ID_SUCCESS:
      let currentResult = state;

      if (!currentResult || !Array.isArray(currentResult)) {
        currentResult = [];
      }

      const actionSafe = action as T.FetchUserPersonalIdSuccessAction;
      const newResult = [...currentResult];

      const index = currentResult.findIndex(
        personalId => personalId?.id === actionSafe.payload.id,
      );

      if (index !== -1) {
        newResult[index] = actionSafe.payload;
      } else {
        newResult.push(actionSafe.payload);
      }

      return newResult;

    default:
      return state;
  }
};

const defaultPersonalId = createReducerWithResult(
  'FETCH_USER_DEFAULT_PERSONAL_ID',
  INITIAL_STATE.defaultPersonalId,
  actionTypes,
);

export const getError = (
  state: T.UserPersonalIdsState,
): T.UserPersonalIdsState['error'] => state.error;
export const getIsLoading = (
  state: T.UserPersonalIdsState,
): T.UserPersonalIdsState['isLoading'] => state.isLoading;
export const getResult = (
  state: T.UserPersonalIdsState,
): T.UserPersonalIdsState['result'] => state.result;
export const getDefaultPersonalId = (
  state: T.UserPersonalIdsState,
): T.UserPersonalIdsState['defaultPersonalId'] => state.defaultPersonalId;

const reducer = combineReducers({
  isLoading,
  result,
  error,
  defaultPersonalId,
});

/**
 * Reducer for personal ids state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const personalIdsReducer: Reducer<T.UserPersonalIdsState> = (state, action) => {
  if (action.type === actionTypes.RESET_USER_PERSONAL_IDS) {
    return INITIAL_STATE;
  }

  return reducer(state, action);
};

export default personalIdsReducer;
