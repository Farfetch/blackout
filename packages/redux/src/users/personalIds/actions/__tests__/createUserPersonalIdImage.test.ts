import * as actionTypes from '../../actionTypes.js';
import {
  config,
  expectedConfig,
  mockPersonalIdResponse,
  personalIdImageData,
  userId,
} from 'tests/__fixtures__/users/index.mjs';
import { createUserPersonalIdImage } from '..//index.js';
import { find } from 'lodash-es';
import { INITIAL_STATE } from '../../../reducer.js';
import { mockStore } from '../../../../../tests/index.js';
import { postUserPersonalIdImage } from '@farfetch/blackout-client';

jest.mock('@farfetch/blackout-client', () => ({
  ...jest.requireActual('@farfetch/blackout-client'),
  postUserPersonalIdImage: jest.fn(),
}));

const usersMockStore = (state = {}) =>
  mockStore({ users: INITIAL_STATE }, state);

describe('createUserPersonalIdImage() action creator', () => {
  let store = usersMockStore();

  beforeEach(() => {
    jest.clearAllMocks();
    store = usersMockStore();
  });

  it('should create the correct actions for when the create user personal id image procedure fails', async () => {
    const expectedError = new Error('create user personal id image error');

    (postUserPersonalIdImage as jest.Mock).mockRejectedValueOnce(expectedError);

    await expect(
      async () =>
        await createUserPersonalIdImage(
          userId,
          personalIdImageData,
          config,
        )(store.dispatch),
    ).rejects.toThrow(expectedError);

    expect(postUserPersonalIdImage).toHaveBeenCalledTimes(1);
    expect(postUserPersonalIdImage).toHaveBeenCalledWith(
      userId,
      personalIdImageData,
      expectedConfig,
    );
    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        { type: actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_REQUEST },
        {
          type: actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_FAILURE,
          payload: { error: expectedError },
        },
      ]),
    );
  });

  it('should create the correct actions for when the create user personal id image procedure is successful', async () => {
    (postUserPersonalIdImage as jest.Mock).mockResolvedValueOnce(
      mockPersonalIdResponse,
    );
    await createUserPersonalIdImage(
      userId,
      personalIdImageData,
      config,
    )(store.dispatch);

    const actionResults = store.getActions();

    expect(postUserPersonalIdImage).toHaveBeenCalledTimes(1);
    expect(postUserPersonalIdImage).toHaveBeenCalledWith(
      userId,
      personalIdImageData,
      expectedConfig,
    );

    expect(actionResults).toMatchObject([
      { type: actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_REQUEST },
      {
        payload: mockPersonalIdResponse,
        type: actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_SUCCESS,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.CREATE_USER_PERSONAL_ID_IMAGE_SUCCESS,
      }),
    ).toMatchSnapshot('create user personal id image success payload');
  });
});
