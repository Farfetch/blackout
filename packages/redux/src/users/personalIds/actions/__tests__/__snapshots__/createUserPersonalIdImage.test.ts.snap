// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createUserPersonalIdImage() action creator should create the correct actions for when the create user personal id image procedure is successful: create user personal id image success payload 1`] = `
Object {
  "payload": Object {
    "backImageId": "",
    "expiryDate": "",
    "frontImageId": "",
    "id": "",
    "isDefault": true,
    "maskedIdNumber": "",
    "maskedName": "",
    "verifyLevel": 1,
  },
  "type": "@farfetch/blackout-redux/CREATE_USER_PERSONAL_ID_IMAGE_SUCCESS",
}
`;
